File: backend-api/schemas/group_invitation.py
from pydantic import BaseModel, Field, ConfigDict
from typing import Optional, List
from datetime import datetime
from enum import Enum

class GroupInvitationStatus(str, Enum):
    PENDING = "pending"
    ACCEPTED = "accepted"
    DECLINED = "declined"
    REVOKED = "revoked"
    EXPIRED = "expired"

class GroupInvitationBase(BaseModel):
    group_id: int
    email: str = Field(..., description="Email of the invited user")
    role: str = Field("member", description="Role to be assigned when accepted")

class GroupInvitationCreate(GroupInvitationBase):
    pass

class GroupInvitationUpdate(BaseModel):
    role: Optional[str] = Field(None, description="Role to be assigned when accepted")

class GroupInvitationAccept(BaseModel):
    token: str = Field(..., description="Invitation token")

class GroupInvitationDecline(BaseModel):
    token: str = Field(..., description="Invitation token")

class GroupInvitationRevoke(BaseModel):
    id: int = Field(..., description="Invitation ID")

class GroupInvitation(GroupInvitationBase):
    id: int
    invited_by_user_id: int
    token: str
    expires_at: datetime
    accepted_at: Optional[datetime] = None
    declined_at: Optional[datetime] = None
    revoked_at: Optional[datetime] = None
    created_at: datetime
    status: GroupInvitationStatus

    @property
    def is_expired(self) -> bool:
        return datetime.utcnow() > self.expires_at

    model_config = ConfigDict(from_attributes=True)

class GroupInvitationListResponse(BaseModel):
    items: List[GroupInvitation]
    total: int

File: frontend/src/types/groupInvitation.ts
export interface GroupInvitationBase {
  group_id: number;
  email: string;
  role: string;
}

export interface GroupInvitationCreate extends GroupInvitationBase {
}

export interface GroupInvitationUpdate {
  role?: string;
}

export interface GroupInvitationAccept {
  token: string;
}

export interface GroupInvitationDecline {
  token: string;
}

export interface GroupInvitationRevoke {
  id: number;
}

export interface GroupInvitation extends GroupInvitationBase {
  id: number;
  invited_by_user_id: number;
  token: string;
  expires_at: string;
  accepted_at: string | null;
  declined_at: string | null;
  revoked_at: string | null;
  created_at: string;
  status: 'pending' | 'accepted' | 'declined' | 'revoked' | 'expired';
}

export interface GroupInvitationListResponse {
  items: GroupInvitation[];
  total: number;
}